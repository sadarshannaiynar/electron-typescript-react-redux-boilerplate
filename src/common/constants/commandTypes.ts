export const EXEC_LS = 'EXEC_LS';