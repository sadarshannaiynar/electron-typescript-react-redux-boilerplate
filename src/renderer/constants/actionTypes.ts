export const INCREMENT_VALUE = 'INCREMENT_VALUE';
export const DECREMENT_VALUE = 'DECREMENT_VALUE';
export const SET_VALUE = 'SET_VALUE';
export const EXEC_LS_SUCCESS = 'EXEC_LS_SUCCESS';
export const EXEC_LS_ERROR = 'EXEC_LS_ERROR';
