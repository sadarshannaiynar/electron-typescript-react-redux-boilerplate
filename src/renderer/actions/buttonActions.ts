import { IpcMessageEvent, IpcRenderer } from 'electron';
import { Dispatch } from 'redux';
import { action } from 'typesafe-actions';
import { EXEC_LS } from '../../common/constants/commandTypes';
import { DECREMENT_VALUE, EXEC_LS_ERROR, EXEC_LS_SUCCESS, INCREMENT_VALUE, SET_VALUE } from '../constants/actionTypes';

export interface IEventOutput {
  log: string;
}

export const incrementAction = () => action(INCREMENT_VALUE);

export const decrementAction = () => action(DECREMENT_VALUE);

export const setValue = (value: number) => action(SET_VALUE, value);

export const startExecuteLs = () => action(EXEC_LS);

export const successExecuteLs = (output: string) => action(EXEC_LS_SUCCESS, output);

export const errorExecuteLs = (error: string) => action(EXEC_LS_ERROR, error);

export const executeLs = (ipcRenderer: IpcRenderer): any => (
  (dispatch: Dispatch) => {
    ipcRenderer.send(EXEC_LS);
    dispatch(startExecuteLs());

    ipcRenderer.once(`success-${EXEC_LS}`, (event: IpcMessageEvent, { log }: IEventOutput) => {
      dispatch(successExecuteLs(log));
      ipcRenderer.removeAllListeners(`error-${EXEC_LS}`);
    });

    ipcRenderer.once(`error-${EXEC_LS}`, (event: IpcMessageEvent, { log }: IEventOutput) => {
      dispatch(errorExecuteLs(log));
      ipcRenderer.removeAllListeners(`success-${EXEC_LS}`);
    });
  }
);
