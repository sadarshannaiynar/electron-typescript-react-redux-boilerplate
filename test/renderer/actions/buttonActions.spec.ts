import { IpcMessageEvent } from 'electron';
import reduxMockStore from 'redux-mock-store';
import reduxThunk from 'redux-thunk';
import { EXEC_LS } from '../../../src/common/constants/commandTypes';
import * as actions from '../../../src/renderer/actions/buttonActions';

// tslint:disable-next-line:no-var-requires
const electronIpcMock = require('electron-ipc-mock')();

describe('buttonActions Tests', () => {
  describe('Actions Tests', () => {
    test('should return the increment action', () => {
      expect(actions.incrementAction()).toMatchSnapshot();
    });

    test('should return the decrement action', () => {
      expect(actions.decrementAction()).toMatchSnapshot();
    });

    test('should return the setValue action', () => {
      expect(actions.setValue(0)).toMatchSnapshot();
    });

    test('should return the startExecuteLs action', () => {
      expect(actions.startExecuteLs()).toMatchSnapshot();
    });

    test('should return the successExecuteLs action', () => {
      expect(actions.successExecuteLs('Success')).toMatchSnapshot();
    });

    test('should return the errorExecuteLs action', () => {
      expect(actions.errorExecuteLs('Error')).toMatchSnapshot();
    });
  });

  describe('Store Dispatch Tests', () => {
    const mockStore = reduxMockStore([reduxThunk]);

    test('should update the store on increment action', () => {
      const store = mockStore({});
      store.dispatch(actions.incrementAction());
      expect(store.getActions()[0]).toMatchSnapshot();
    });

    test('should update the store on decrement action', () => {
      const store = mockStore({});
      store.dispatch(actions.decrementAction());
      expect(store.getActions()[0]).toMatchSnapshot();
    });

    test('should update the store on setValue action', () => {
      const store = mockStore({});
      store.dispatch(actions.setValue(0));
      expect(store.getActions()[0]).toMatchSnapshot();
    });

    test('should update the store on executeLs action with success reply', (done) => {
      const store = mockStore({});
      const { ipcMain, ipcRenderer } = electronIpcMock;
      ipcMain.on(EXEC_LS, (event: IpcMessageEvent) => {
        event.sender.send(`success-${EXEC_LS}`, { log: 'Output' });
        ipcMain.removeAllListeners(EXEC_LS);
      });

      store.dispatch(actions.executeLs(ipcRenderer));
      expect(store.getActions()[0]).toMatchObject(actions.startExecuteLs());

      ipcRenderer.on(`success-${EXEC_LS}`, () => {
        expect(store.getActions()[1]).toMatchObject(actions.successExecuteLs('Output'));
        expect(store.getActions()).toMatchSnapshot();
        ipcRenderer.removeAllListeners(`success-${EXEC_LS}`);
        done();
      });
    });

    test('should update the store on executeLs action with error reply', (done) => {
      const store = mockStore({});
      const { ipcMain, ipcRenderer } = electronIpcMock;
      ipcMain.on(EXEC_LS, (event: IpcMessageEvent) => {
        event.sender.send(`error-${EXEC_LS}`, { log: 'Error' });
        ipcMain.removeAllListeners(EXEC_LS);
      });

      store.dispatch(actions.executeLs(ipcRenderer));
      expect(store.getActions()[0]).toMatchObject(actions.startExecuteLs());

      ipcRenderer.on(`error-${EXEC_LS}`, () => {
        expect(store.getActions()[1]).toMatchObject(actions.errorExecuteLs('Error'));
        expect(store.getActions()).toMatchSnapshot();
        ipcRenderer.removeAllListeners(`error-${EXEC_LS}`);
        done();
      });
    });
  });
});
