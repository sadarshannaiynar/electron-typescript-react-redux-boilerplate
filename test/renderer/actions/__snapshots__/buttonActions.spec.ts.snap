// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buttonActions Tests Actions Tests should return the decrement action 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "DECREMENT_VALUE",
}
`;

exports[`buttonActions Tests Actions Tests should return the errorExecuteLs action 1`] = `
Object {
  "meta": undefined,
  "payload": "Error",
  "type": "EXEC_LS_ERROR",
}
`;

exports[`buttonActions Tests Actions Tests should return the increment action 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "INCREMENT_VALUE",
}
`;

exports[`buttonActions Tests Actions Tests should return the setValue action 1`] = `
Object {
  "meta": undefined,
  "payload": 0,
  "type": "SET_VALUE",
}
`;

exports[`buttonActions Tests Actions Tests should return the startExecuteLs action 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "EXEC_LS",
}
`;

exports[`buttonActions Tests Actions Tests should return the successExecuteLs action 1`] = `
Object {
  "meta": undefined,
  "payload": "Success",
  "type": "EXEC_LS_SUCCESS",
}
`;

exports[`buttonActions Tests Store Dispatch Tests should update the store on decrement action 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "DECREMENT_VALUE",
}
`;

exports[`buttonActions Tests Store Dispatch Tests should update the store on executeLs action with error reply 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "EXEC_LS",
  },
  Object {
    "meta": undefined,
    "payload": "Error",
    "type": "EXEC_LS_ERROR",
  },
]
`;

exports[`buttonActions Tests Store Dispatch Tests should update the store on executeLs action with success reply 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "EXEC_LS",
  },
  Object {
    "meta": undefined,
    "payload": "Output",
    "type": "EXEC_LS_SUCCESS",
  },
]
`;

exports[`buttonActions Tests Store Dispatch Tests should update the store on increment action 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "INCREMENT_VALUE",
}
`;

exports[`buttonActions Tests Store Dispatch Tests should update the store on setValue action 1`] = `
Object {
  "meta": undefined,
  "payload": 0,
  "type": "SET_VALUE",
}
`;
